### PR TITLE
Elastic AllReduce controller for TensorFlow 1.x

### PR DESCRIPTION
**Pull request by workingloong**

### CODE REVIEW

File: elasticdl/python/tests/allreduce_trainer_test.py [hunk 1]
```diff
@@ -163,8 +163,8 @@ def test_elastic_run(self):
         controller.set_broadcast_model(model)
         controller.set_broadcast_optimizer(optimizer)
         controller.broadcast()
-        result = controller.train_one_batch_with_retries(self.train)
         controller.restore()
+        result = controller.train_one_batch_with_retries(self.train)
```
Comment (brightcoder01) [inline elasticdl/python/tests/allreduce_trainer_test.py]: Why do we move `train_one_batch_with_retries` after `restore`?

File: elasticdl/python/allreduce/tensorflow_controller.py [hunk 2]
```diff
@@ -74,3 +75,44 @@ def train_one_batch_with_retries(self, func, *args, **kwargs):
...
+                return result
+            except UnknownError as e:
+                logger.warning(
+                    "Failed to perform allreduce operation on "
+                    "the gradients. Retrying..."
+                )
+                # Those error message show that the communication
+                # to merge gradient fails and we can rebuild the
+                # communication.
+                if (
```
Comment (brightcoder01) [inline elasticdl/python/allreduce/tensorflow_controller.py:112]: How about the `HorovodUpdateError` and `HorovodInternalError`?

File: elasticdl/python/allreduce/tensorflow_controller.py [hunk 3]
```diff
@@ -74,3 +75,44 @@ def train_one_batch_with_retries(self, func, *args, **kwargs):
                     self._rendezvous_manager.init_horovod_if_needed()
         self.data_shard_service.report_batch_done()
         return result
+
+
+class TensorFlowV1AllReduceController(AllReduceController):
+    """The controller is responsible for elastic training of
+    TensorFlow eager execution using AllReduce.
```
Comment (brightcoder01) [inline elasticdl/python/allreduce/tensorflow_controller.py]: eager -> graph

File: elasticdl/python/tests/allreduce_trainer_test.py (hunk 1 above)
Comment (workingloong) [reply elasticdl/python/tests/allreduce_trainer_test.py]: This is a mistake in the PR and I will remove the modification.

File: elasticdl/python/allreduce/tensorflow_controller.py (hunk 2 above)
Comment (workingloong) [reply elasticdl/python/allreduce/tensorflow_controller.py:112]: `HorovodUpdateError` and `HorovodInternalError` are the errors in Horovod not TensorFlow. Horovod will convert the `UnknownError` in TensorFlow to those errors.
https://github.com/horovod/horovod/blob/e94d8ea0dff8c2b45698cfe4fabb2e6553d0b9a8/horovod/tensorflow/elastic.py#L53-L61

Here, we get the error from TensorFlow, not Horovod. So we can not use `HorovodUpdateError` and `HorovodInternalError`.

**Review by brightcoder01 — APPROVED** [review]

**Pull request merged**